semaphore.go

```diff
@@ -0,0 +1,58 @@
+package influxdb
+
+import (
+	"context"
+	"time"
+)
+
+// DefaultLeaseTTL is used when a specific lease TTL is not requested.
+const DefaultLeaseTTL = time.Minute
+
+// A Semaphore provides an API for requesting ownership of an expirable semaphore.
+//
+// Acquired semaphores have an expiration period. If they're not released or extended
+// during this period then they will expire and ownership of the semaphore will
+// be lost.
+type Semaphore interface {
+	// TODO(edd): add Acquire and AcquireTTL when needed. These should block.
+
+	// TryAcquire attempts to acquire ownership of the semaphore. TryAcquire
+	// must not block. Failure to get ownership of the semaphore should be
+	// signalled to the caller via the return of a nil Lease.
+	TryAcquire(context.Context) (Lease, error)
+
+	// TryAcquireTTL is similar to TryAcquire, but a specific TTL is provided.
+	TryAcquireTTL(ctx context.Context, ttl time.Duration) (Lease, error)
+}
```

Comment:
Did you not like the name, `Lessor`? Acquiring a lease from a semaphore seems odd to me personally.

Reply:
If anything `Lease` could be renamed to something else, but not sure what. I kind of worried about this. Usually you wouldn't return anything aside from maybe a `bool` indicating you have ownership, but in this case I wanted something that can be extended and revoked early. It made more sense to return that complex logic in a value. Essentially the semaphore is granting leased ownership over some resource (since leases have a fixed duration).